From df2e6366e86a46acb6353b523c2fbe707daba1a7 Mon Sep 17 00:00:00 2001
From: Luciano Coelho <coelho@ti.com>
Date: Wed, 1 Jun 2011 14:55:41 +0300
Subject: [PATCH 75/75] mac80211: leave dynamic PS when receiving data

This patch triggers the dynamic PS mechanism to leave PS mode when a
packet is received by queuing a dynamic PS disable work.

This is needed in order to leave PS mode when packets are only
received and not transmitted (ie. UDP RX).  Not leaving PS mode when
receiving packets may cause some interoperability problems with
certain APs.

We also leave PS when a multicast packet is received, but not
broadcast.

Signed-off-by: Luciano Coelho <coelho@ti.com>
---
 net/mac80211/rx.c |    9 ++++++---
 1 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 7fa8c6b..e8e67dc 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -1964,12 +1964,15 @@ ieee80211_rx_h_data(struct ieee80211_rx_data *rx)
 	dev->stats.rx_bytes += rx->skb->len;
 
 	if (local->ps_sdata && local->hw.conf.dynamic_ps_timeout > 0 &&
-	    !is_multicast_ether_addr(
+	    !is_broadcast_ether_addr(
 		    ((struct ethhdr *)rx->skb->data)->h_dest) &&
 	    (!local->scanning &&
 	     !test_bit(SDATA_STATE_OFFCHANNEL, &sdata->state))) {
-			mod_timer(&local->dynamic_ps_timer, jiffies +
-			 msecs_to_jiffies(local->hw.conf.dynamic_ps_timeout));
+		ieee80211_queue_work(&local->hw,
+				     &local->dynamic_ps_disable_work);
+
+		mod_timer(&local->dynamic_ps_timer, jiffies +
+			  msecs_to_jiffies(local->hw.conf.dynamic_ps_timeout));
 	}
 
 	ieee80211_deliver_skb(rx);
-- 
1.7.0.4

